core/src/subgraph/loader.rs

```diff
@@ -4,207 +4,71 @@ use std::time::Instant;
-                        "Not all dynamic data sources of deployment `{}` are objects",
-                        deployment_id
-                    ))
-                }
-            }
-            _ => Err(format_err!(
-                "Dynamic data sources of deployment `{}` are not a list",
-                deployment_id
-            )),
-        }?;
-
-        // Parse the raw data sources into typed entities
-        let entities = values.iter().try_fold(vec![], |mut entities, value| {
-            entities.push(UnresolvedDataSource::try_from_value(value)?);
-            Ok(entities)
-        });
-
-        entities.map_err(|e: Error| {
-            format_err!(
-                "Failed to parse dynamic data source entities of deployment `{}`: {}",
-                deployment_id,
-                e
-            )
-        })
-    }
-
-    async fn resolve_data_sources(
-        &self,
-        unresolved_data_sources: Vec<UnresolvedDataSource>,
-        logger: &Logger,
-    ) -> Result<Vec<DataSource>, Error> {
-        // Resolve the data sources and return them
-        let mut result = Vec::new();
-        for item in unresolved_data_sources.into_iter() {
-            let resolved = item.resolve(&*self.link_resolver, logger).await?;
-            result.push(resolved);
-        }
-        Ok(result)
+    pub fn new(store: Arc<S>) -> Self {
+        Self { store }
     }
 }
 
 #[async_trait]
-impl<L, Q, S> DataSourceLoaderTrait for DataSourceLoader<L, Q, S>
+impl<S> DataSourceLoaderTrait for DataSourceLoader<S>
 where
-    L: LinkResolver,
-    Q: GraphQlRunner,
     S: Store + SubgraphDeploymentStore,
 {
     async fn load_dynamic_data_sources(
         &self,
         deployment_id: SubgraphDeploymentId,
         logger: Logger,
+        manifest: SubgraphManifest,
     ) -> Result<Vec<DataSource>, Error> {
         let start_time = Instant::now();
 
+        let template_map: HashMap<&str, &DataSourceTemplate> = HashMap::from_iter(
+            manifest
+                .templates
+                .iter()
+                .map(|template| (template.name.as_str(), template)),
+        );
         let mut data_sources = vec![];
 
-        loop {
-            let skip = data_sources.len() as i32;
-            let query = self.dynamic_data_sources_query(&deployment_id, skip)?;
-            let query_result = self
-                .graphql_runner
-                .cheap_clone()
-                .query_metadata(query)
-                .await?;
-            let unresolved_data_sources =
-                self.parse_data_sources(&deployment_id, query_result.as_ref().clone())?;
-            let next_data_sources = self
-                .resolve_data_sources(unresolved_data_sources, &logger)
-                .await?;
-
-            if next_data_sources.is_empty() {
-                break;
-            }
-
-            data_sources.extend(next_data_sources);
+        for stored in self.store.load_dynamic_data_sources(&deployment_id)? {
```

Comment:
Not great that this calls a blocking function from an `async` fn. Could we leverage `with_entity_conn` to make this an async store fn?

Reply:
I just tried, and ran into the issue of going back and forth between `anyhow::Error` and `StoreError`. Can we do an overhaul of errors from the store and asyncifying this after we are done with the sharding PR I am going to open in a sec?

Reply:
Fine by me.